src/validator.cpp

```diff
@@ -1635,6 +1708,530 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasOneOrTwoMathmlChildren(const XmlNodePtr &node,
+                                                         const ComponentPtr &component)
+{
+    auto childCount = mathmlChildCount(node);
+
+    if ((childCount != 1) && (childCount != 2)) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without one or two MathML children.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+void Validator::ValidatorImpl::validateMathMLElementsChildrenAndSiblings(const XmlNodePtr &node,
+                                                                         const ComponentPtr &component)
+{
+    // Check the current node against the MathML elements listed in
+    // supportedMathMLElements.
+    // Note: we don't have to do anything for 'true', 'false, 'exponentiale',
+    //       'pi', 'infinity', and 'notanumber' since they must be empty and if
+    //       they are not then this will be caught by the DTD.
+
+    // Basic content elements.
+
+    if (node->isMathmlElement("apply")) {
+        if (hasAtLeastOneMathmlChild(node, component)) {
+            for (size_t i = 0, iMax = mathmlChildCount(node); i < iMax; ++i) {
+                validateMathMLElementsChildrenAndSiblings(mathmlChildNode(node, i), component);
+            }
+        }
+
+        // Assignment, and relational and logical operators.
+
+    } else if (node->isMathmlElement("eq")
+               || node->isMathmlElement("neq")
+               || node->isMathmlElement("lt")
+               || node->isMathmlElement("leq")
+               || node->isMathmlElement("gt")
+               || node->isMathmlElement("geq")) {
+        auto parentNode = node->parent();
+
+        hasTwoMathmlSiblings(parentNode, node, component)
+            && isFirstMathmlChild(parentNode, node, component);
+    } else if (node->isMathmlElement("and")
+               || node->isMathmlElement("or")
+               || node->isMathmlElement("xor")) {
+        auto parentNode = node->parent();
+
+        hasAtLeastTwoMathmlSiblings(parentNode, node, component)
+            && isFirstMathmlChild(parentNode, node, component);
+    } else if (node->isMathmlElement("not")) {
+        auto parentNode = node->parent();
+
+        hasOneMathmlSibling(parentNode, node, component)
+            && isFirstMathmlChild(parentNode, node, component);
+
+        // Arithmetic operators.
+
+    } else if (node->isMathmlElement("plus")
+               || node->isMathmlElement("minus")) {
+        auto parentNode = node->parent();
+
+        hasAtLeastOneMathmlSibling(parentNode, node, component)
+            && isFirstMathmlChild(parentNode, node, component);
+    } else if (node->isMathmlElement("times")
+               || node->isMathmlElement("divide")) {
+        auto parentNode = node->parent();
```

Comment:
divide is a binary operator (times is n-ary)

Reply:
Damn it! I made the same mistake for the `minus` operator which is either a unary or a binary operator (see 5a93957f).

ac71b283